### PR TITLE
Make citus_stat_tenants work with schema-based tenants.

### PR DESCRIPTION
**Pull request by gokhangulbiz**
DESCRIPTION: Enabling citus_stat_tenants to support schema-based tenants.

This pull request modifies the existing logic to enable tenant monitoring with schema-based tenants. The changes made are as follows:

- If a query has a partitionKeyValue (which serves as a tenant key/identifier for distributed tables), Citus annotates the query with both the partitionKeyValue and colocationId. This allows for accurate tracking of the query. 
- If a query does not have a partitionKeyValue, but its colocationId belongs to a distributed schema, Citus annotates the query with only the colocationId. The tenant monitor can then easily look up the schema to determine if it's a distributed schema and make a decision on whether to track the query.

### CODE REVIEW

File: src/backend/distributed/utils/citus_stat_tenants.c [hunk 1]
```diff
@@ -236,26 +257,51 @@ AttributeTask(char *tenantId, int colocationId, CmdType commandType)
...
  * AnnotateQuery annotates the query with tenant attributes.
+ * if the query has a partition key, we annotate it with the partition key value and colocationId
+ * if the query doesn't have a partition key and if it's a schema-based tenant, we annotate it with the colocationId only.
  */
 char *
 AnnotateQuery(char *queryString, Const *partitionKeyValue, int colocationId)
 {
-	if (StatTenantsTrack == STAT_TENANTS_TRACK_NONE || partitionKeyValue == NULL)
+	if (StatTenantsTrack == STAT_TENANTS_TRACK_NONE || colocationId ==
+		INVALID_COLOCATION_ID)
```
Comment (JelteF) [inline src/backend/distributed/utils/citus_stat_tenants.c:267]: nit: let's put this in a more reasonable place (it's one of the things citus_indent messes up sometimes).
```suggestion
	if (StatTenantsTrack == STAT_TENANTS_TRACK_NONE || 
	    colocationId == INVALID_COLOCATION_ID)
```

File: src/backend/distributed/executor/local_executor.c [hunk 2]
```diff
@@ -382,20 +382,16 @@ ExecuteLocalTaskListExtended(List *taskList,
 
 /*
  * SetColocationIdAndPartitionKeyValueForTasks sets colocationId and partitionKeyValue
- * for the tasks in the taskList if workerJob has a colocationId and partitionKeyValue.
+ * for the tasks in the taskList.
  */
 static void
 SetColocationIdAndPartitionKeyValueForTasks(List *taskList, Job *workerJob)
 {
-	if (workerJob->colocationId != 0 &&
```
Comment (JelteF) [inline src/backend/distributed/executor/local_executor.c:390]: I think we can keep the `workerJob->colocationId != 0` check.

File: src/backend/distributed/utils/citus_stat_tenants.c [hunk 3]
```diff
@@ -221,41 +233,86 @@ void
...
+	 * we try to get the tenantId from the colocationId to lookup schema name and use it as a tenantId
+	 */
+	if (tenantId == NULL)
+	{
+		if (!IsTenantSchemaColocationGroup(colocationId))
+		{
+			return;
+		}
+
+	}
```
Comment (JelteF) [inline src/backend/distributed/utils/citus_stat_tenants.c:252]: I don't think this check is needed anymore. We only annotate the query if with tenantId=NULL if IsTenantSchemaColocationGroup returned true. So no need to re-check it here again.

File: src/test/regress/sql/citus_stat_tenants.sql [hunk 4]
```diff
@@ -311,5 +311,56 @@ SELECT count(*)>=0 FROM select_from_dist_tbl_text_view WHERE a = U&'\0061\0308bc
 
 SELECT tenant_attribute, query_count_in_this_period FROM citus_stat_tenants;
 
+-- schema based tenants
```
Comment (JelteF) [inline src/test/regress/sql/citus_stat_tenants.sql:314]: We should add a simple test to see that queries are not tracked when the query is on a single shard distributed table, which is not part of a tenant schema `create_distributed_table('t1', NULL)`. A very simple test is fine, just like the one we have for reference tables now.

File: src/backend/distributed/executor/local_executor.c [hunk 5]
```diff
@@ -382,13 +382,12 @@ ExecuteLocalTaskListExtended(List *taskList,
...
-		workerJob->partitionKeyValue != NULL)
+	if (workerJob->colocationId != 0)
```
Comment (JelteF) [inline src/backend/distributed/executor/local_executor.c:390]: nit: For readability let's use `INVALID_COLOCATION_ID`:

```suggestion
	if (workerJob->colocationId != INVALID_COLOCATION_ID)
```

I noticed we were comparing to this named constant in other places.

File: src/backend/distributed/utils/citus_stat_tenants.c [hunk 6]
```diff
@@ -683,7 +740,8 @@ FindTenantStats(MultiTenantMonitor *monitor)
 	for (int i = 0; i < monitor->tenantCount; i++)
 	{
 		TenantStats *tenantStats = &monitor->tenants[i];
-		if (strcmp(tenantStats->tenantAttribute, AttributeToTenant) == 0 &&
+
+		if (strncmp(tenantStats->tenantAttribute, AttributeToTenant, MAX_TENANT_ATTRIBUTE_LENGTH) == 0 &&
```
Comment (JelteF) [inline src/backend/distributed/utils/citus_stat_tenants.c:744]: Why was this change necessary?

Comment (gokhangulbiz) [reply src/backend/distributed/utils/citus_stat_tenants.c]: I don't have a solid reason. I was just curious about its impact on performance. In any case, it's no longer necessary.

File: src/backend/distributed/utils/citus_stat_tenants.c (hunk 3 above)
Comment (gokhangulbiz) [reply src/backend/distributed/utils/citus_stat_tenants.c]: You're right, except for locally executed queries (ex: queries to reference tables) that directly follow the AttributeTask(..) flow without being annotated. We can move this check to [local_executor](https://github.com/citusdata/citus/blob/gokhangulbiz/schema-based-tenant-monitoring/src/backend/distributed/executor/local_executor.c#L692). Let me know your thoughts.

**Review by JelteF — APPROVED** [review]
